qurro/tests/test_qarcoal.py

```diff
@@ -155,24 +220,223 @@ def test_samples_to_use(self, get_mp_data):
+        table = get_testing_data.table.to_dataframe()
+        taxonomy = get_testing_data.taxonomy
+        taxonomy_miss = taxonomy.iloc[:-1, :]  # remove F5 from taxonomy
+        num_df, denom_df = filter_and_join_taxonomy(
+            get_testing_data.table.to_dataframe(),
+            taxonomy_miss,
+            "Char",
+            "saur",
+        )
+
+        # F5 dropped out
+        assert "F5" not in num_df.index
+
+        num_features = ["F3", "F4"]
+        denom_features = ["F1", "F2"]
+        sample_order = ["S{}".format(i) for i in range(5)]
+
+        table_num_taxon_filt = pd.DataFrame(table.loc[num_features])
+        table_denom_taxon_filt = pd.DataFrame(table.loc[denom_features])
+
+        num_df_taxon_filt = num_df.loc[num_features]
+        denom_df_taxon_filt = denom_df.loc[denom_features]
+
+        # set operations in Qarcoal change order of samples
+        table_num_taxon_filt = table_num_taxon_filt[sample_order]
+        table_denom_taxon_filt = table_denom_taxon_filt[sample_order]
+        num_df_taxon_filt = num_df_taxon_filt[sample_order]
+        denom_df_taxon_filt = denom_df_taxon_filt[sample_order]
+
+        # test that num/denom df accurately extract values from table
+        assert table_num_taxon_filt.equals(num_df_taxon_filt)
+        assert table_denom_taxon_filt.equals(denom_df_taxon_filt)
+
+    def test_feat_table_missing_features(self, get_testing_data):
+        """Feature table missing features that are present in taxonomy"""
+        table = get_testing_data.table.to_dataframe()
+        table = table.iloc[:-1, :]  # remove F5 from feature table
+        num_df, denom_df = filter_and_join_taxonomy(
+            table,
+            get_testing_data.taxonomy,
+            "Char",
+            "saur",
+        )
+
+        # F5 dropped out
+        assert "F5" not in num_df.index
+
+        num_features = ["F3", "F4"]
+        denom_features = ["F1", "F2"]
+        sample_order = ["S{}".format(i) for i in range(5)]
+
+        table_num_taxon_filt = pd.DataFrame(table.loc[num_features])
+        table_denom_taxon_filt = pd.DataFrame(table.loc[denom_features])
+
+        num_df_taxon_filt = num_df.loc[num_features]
+        denom_df_taxon_filt = denom_df.loc[denom_features]
+
+        # set operations in Qarcoal change order of samples
+        table_num_taxon_filt = table_num_taxon_filt[sample_order]
+        table_denom_taxon_filt = table_denom_taxon_filt[sample_order]
+        num_df_taxon_filt = num_df_taxon_filt[sample_order]
+        denom_df_taxon_filt = denom_df_taxon_filt[sample_order]
+
+        # test that num/denom df accurately extract values from table
+        assert table_num_taxon_filt.equals(num_df_taxon_filt)
+        assert table_denom_taxon_filt.equals(denom_df_taxon_filt)
+
+    def test_overlapping_columns(self, get_testing_data):
+        """Feature table and taxonomy have overlapping column(s)"""
```

Comment:
As I understand it, this case isn't explicitly handled since `filter_and_join_taxonomy()` removes all taxonomy columns that aren't `Taxon`, right? It might be worth adding a note specifying that here.

Reply:
I think I may be misunderstanding your comment. The idea behind this test was that if there are overlapping columns, the removing of all non-`Taxon` columns would account for it (i.e. the removal *is* the test). (Happy to clarify/discuss in-person if needed :grinning:)